GatherBuddy/Plugin/OceanUptime.cs

```diff
@@ -0,0 +1,224 @@
+            new TimedRoute(OceanRoute.NorthernStraitofMerlthor, OceanTime.Night),
+            new TimedRoute(OceanRoute.RhotanoSea, OceanTime.Night),
+
+            new TimedRoute(OceanRoute.BloodbrineSea, OceanTime.Day),
+            new TimedRoute(OceanRoute.RothlytSound, OceanTime.Day),
+            new TimedRoute(OceanRoute.NorthernStraitofMerlthor, OceanTime.Day),
+            new TimedRoute(OceanRoute.RhotanoSea, OceanTime.Day),
+        };
+
+        var loop = new TimedRoute[144];
+
+        int outputIdx = 0;
+
+        // The full fixed route first (-1), and then the fixed route
+        // but removing one item (in order) each time, for a total of
+        // 12 + 11 * 12 = 144 items in the loop.
+        for (int skipIdx = -1; skipIdx < fixedRouteOrder.Length; skipIdx++)
+        {
+            for (int routeIdx = 0; routeIdx < fixedRouteOrder.Length; routeIdx++)
+            {
+                if (routeIdx == skipIdx)
+                    continue;
+                loop[outputIdx] = fixedRouteOrder[routeIdx];
+                outputIdx++;
+            }
+        }
+
+        Debug.Assert(outputIdx == 144);
+        return loop;
+    }
+
+
+    // Returns the current/next TimeInterval from a utc timestamp for this ocean fish.
+    public static TimeInterval GetOceanUptime(Fish fish, TimeStamp utcNow)
+    {
+        if (!fish.OceanFish || fish.OceanTimes.Length == 0)
+            return TimeInterval.Always;
+
+        // The offset in milliseconds from when the current loop started to the current time.
+        long loopOffsetMilliseconds = (utcNow.Time - LoopTimestampEpoch) % LoopDurationMilliseconds;
+        // The index into RouteLoop for the given timestamp.
+        var loopIdx = loopOffsetMilliseconds / TwoHoursInMilliseconds;
+        // The timestamp that this loop started at.
+        long loopStartTimestamp = utcNow.Time - loopOffsetMilliseconds;
+
+        var spotIds = fish.FishingSpots.Select(spot => spot.Id);
+
+        // Every stop/time combination exists in at least two adjacent fixed routes (since it might
+        // be skipped in one of the routes, but can't be skipped twice), so this will need to check
+        // ~22 routes at most.
+        for (int loopCount = 0; loopCount < 25; loopCount++)
+        {
+            var timedRoute = RouteLoop[loopIdx % RouteLoop.Length];
+            foreach (var time in fish.OceanTimes)
+            {
+                if (!Array.Exists(timedRoute.StopsByTime[time], spotId => spotIds.Contains((uint)spotId)))
+                    continue;
+
+                // Arbitrarily say that the window is an hour long rather than two.
+                // Boats can queue up to 15 minutes past the hour, and boats are roughly 25 minutes.
+                // There can be delay for cutscenes/players loading as well.  An hour is safe.
+                var start = loopStartTimestamp + loopIdx * TwoHoursInMilliseconds;
+                var end = start + OneHourInMilliseconds;

```

Comment:
I'm also not sure what the best thing to do here is and I admit this is kind of weird.

Another alternative is to say "are you in the ocean fishing territory" and if not then say the end is 15 minutes in the futurue (since if you're not on a boat by XX:15 you've missed it).